File: hummingbird/iosApp/Hummingbird/DesignSystem/OperationalDuration.swift
import Foundation

enum OperationalDuration {
    static func seconds(_ value: Double, compact: Bool = false) -> String {
        guard value.isFinite else { return "--" }

        let totalSeconds = Int(abs(value).rounded())
        let negative = value < 0 && totalSeconds > 0
        let hours = totalSeconds / 3_600
        let minutes = (totalSeconds % 3_600) / 60
        let seconds = totalSeconds % 60
        var parts: [String] = []

        if hours > 0 {
            parts.append(compact ? "\(hours)h" : "\(hours) hr")
        }
        if hours > 0 || minutes > 0 {
            parts.append(compact ? "\(minutes)m" : "\(minutes) min")
        }
        parts.append(compact ? "\(seconds)s" : "\(seconds) sec")

        return (negative ? "-" : "") + parts.joined(separator: " ")
    }

    static func minutes(_ value: Double, compact: Bool = false) -> String {
        seconds(value * 60, compact: compact)
    }

    static func age(since date: Date, relativeTo reference: Date = Date()) -> String {
        let elapsed = reference.timeIntervalSince(date)
        guard elapsed >= -0.5 else { return "scheduled" }

        return "\(seconds(max(elapsed, 0))) ago"
    }
}

File: hummingbird/androidApp/app/src/main/java/net/acumenus/hummingbird/ui/components/OperationalDuration.kt
package net.acumenus.hummingbird.ui.components

import java.time.Duration
import java.time.Instant
import kotlin.math.abs
import kotlin.math.roundToLong

internal fun formatOperationalSeconds(value: Double, compact: Boolean = false): String {
    if (!value.isFinite()) return "--"

    val totalSeconds = abs(value).roundToLong()
    val negative = value < 0 && totalSeconds > 0
    val hours = totalSeconds / 3_600
    val minutes = (totalSeconds % 3_600) / 60
    val seconds = totalSeconds % 60
    val parts = buildList {
        if (hours > 0) add(if (compact) "${hours}h" else "$hours hr")
        if (hours > 0 || minutes > 0) add(if (compact) "${minutes}m" else "$minutes min")
        add(if (compact) "${seconds}s" else "$seconds sec")
    }

    return (if (negative) "-" else "") + parts.joinToString(" ")
}

internal fun formatOperationalMinutes(value: Number, compact: Boolean = false): String =
    formatOperationalSeconds(value.toDouble() * 60, compact)

internal fun formatOperationalAge(instant: Instant, now: Instant = Instant.now()): String {
    val elapsedMillis = Duration.between(instant, now).toMillis()
    if (elapsedMillis < -500) return "scheduled"

    return "${formatOperationalSeconds(elapsedMillis.coerceAtLeast(0).toDouble() / 1_000)} ago"
}

internal fun formatOperationalOffset(deltaMillis: Long): String {
    val totalSeconds = abs(deltaMillis.toDouble() / 1_000).roundToLong()
    if (totalSeconds == 0L) return "now"

    val sign = if (deltaMillis < 0) "-" else "+"
    return sign + formatOperationalSeconds(totalSeconds.toDouble(), compact = true)
}
